[24.0] Fix history update time after bulk operation

davelopez opened the pull request:
Fixes #18056

For some reason, when adding or removing tags from a dataset the items don't get their `update_time` changed. I forced that in `set_tags_from_list`. I'm not sure if adding or removing a tag can be considered `updating` a dataset, but, somehow, it used to work in the past...

There are some other cases where, after running a bulk operation on all of the items of a History, there is no actual change on the datasets (for example, undeleting a purged dataset) this means the associated history will not trigger a `update_time` change and the bulk operation in the UI will think it is still running even if it finished.

This includes additional checks in the API tests for bulk operations to ensure the history `update_time` gets updated on every operation.


## How to test the changes?
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

davelopez committed:
Add tests to ensure history update_time gets updated

We need this after a bulk operation to ensure the client will catch-up with the results (even if no-op) so it knows the operation ended.

davelopez committed:
Mark the item as updated after altering the associated tags

davelopez committed:
Mark item as updated when undelete had no effect

Usually if the dataset was not deleted or if it was purged and un-deleting does not do anything

mvdbeek commented on lib/galaxy/webapps/galaxy/services/history_contents.py:
```diff
@@ -1510,6 +1513,10 @@ def _change_dbkey(self, item: HistoryItemModel, params: ChangeDbkeyOperationPara
     def _remove_tags(self, trans: ProvidesUserContext, item: HistoryItemModel, params: TagOperationParams):
         trans.tag_handler.remove_tags_from_list(trans.user, item, params.tags, flush=self.flush)
+        # Changing tags does not change the item, but we need to update the history to trigger the completion of the operation
+        item.update()
```
Should those updates instead be moved to the tag handler ? I assume we always want the update time to be altered ?

davelopez replied:
That is what I had in the previous attempt but many tests failed with a 500 status code... I can give it another try and see what it is actually upsetting those tests

davelopez replied:
I think the issue with this is that some of the database models that may call `set_tags_from_list` did not inherit from `UsesCreateAndUpdateTime` like `StoredWorkflow`. `Page`, etc. Does it make more sense to add `UsesCreateAndUpdateTime` to these models instead of skipping the call in https://github.com/galaxyproject/galaxy/pull/18068/commits/983f7c37c7a96dae63ed6271963d3261666a954b?

jdavcs replied:
I don't see why not. As long as they have those two fields and don't have their own `update` method that does something different, then sure.

davelopez replied:
Great! I've done it in https://github.com/galaxyproject/galaxy/pull/18068/commits/51d8743b5f174abb41bffe3259c50a0b19212aee it feels more consistent to me now. Adding or removing tags in most items (including Histories) was not triggering an update of the item.update_time in the past but it feels like it should as Marius suggested.

davelopez committed:
Move update to tag handler

mvdbeek commented on lib/galaxy/webapps/galaxy/services/history_contents.py:
```diff
@@ -1445,6 +1445,9 @@ def _undelete(self, item: HistoryItemModel):
         manager.undelete(item, flush=self.flush)
+        # Again, we need to force an update in the edge case where all selected items are already undeleted
+        # or when the item was purged as undelete will not trigger an update
+        item.update()
```
This should probably also go into the undelete method ? If we didn't do that my followup would be, are you sure we don't need to flush somewhere in the calling context ? But moving this into the undelete method I'm happy to just assume we do the right thing 😅

mvdbeek replied:
Wait, why is it important to change the update time if no change was made ?

davelopez replied:
It is because of the way we check for the completion of the bulk operation in the client... :disappointed: 
When we run the bulk operation the only way we have to tell it is finished (it might run in a background task) is by checking if the history update_time changed.

https://github.com/galaxyproject/galaxy/blob/4e2aaf8e2b88370e5805c02b1dd51bdff2bd0220/client/src/components/History/CurrentHistory/HistoryOperations/DefaultOperations.vue#L48

Maybe we could devise a better way of checking for the completion? I also don't like having to resort to this "hack".

dannon replied:
Can we hand back a handle to the task and poll the tasks API instead?

davelopez replied:
Maybe we can explore that now, it was not an easy option when this was designed as Celery was more "optional/experimental" back then. It will require changes in the API though so I don't know if this will fit as a bug fix for 24.0 :thinking:

davelopez replied:
So, what do you all think about leaving these "hacky" `item.update()` calls here in the bulk operations as a bug fix for 24.0 and trying to follow up in `dev` changing the API to handle asynchronous tasks? Then we can remove them from here.

Refreshing the `history.update_time` is still important when any of the items change after the bulk operation so the history watcher can fetch those changes. But there are cases when a bulk operation legitimately does not alter anything in the items because it doesn't apply (i.e. trying to undelete purged datasets). These cases should not trigger an update of the history, so to avoid the "forever running operation visual issue" we could track the status of the Celery task if the API returned the ID as Dannon suggested.

dannon replied:
I don't mind us fixing these like this for now.  The downside is pretty minimal, and, really, it only applies in a situation when the user really did want to make an update to the history anyway so I don't think anyone's going to care if the update time gets bumped.

That said, the followup and maybe coming up with a clean async client/task interface (maybe using sse's or a websocket as a trial for this?) sounds like a fun project that we should definitely pursue with more runway :)

davelopez committed:
Only call update if it exists

davelopez committed:
Update models to include UsesCreateAndUpdateTime mixin

For the sake of consistency, those models that have `create_time`, `update_time` and `tags` should share the same interface to update the `update_time` when tags are added or removed.
